Python-kivy: init at 1.10.0

vanschelven opened the pull request:
###### Motivation for this change

Kivy is an Open source Python library for rapid development of applications
that make use of innovative user interfaces, such as multi-touch apps.

Most of Kivy's dependencies are optional, so defining a good "default" set of useful dependencies was a bit of a challenge. The present set of dependencies works well for me, but more dependencies, or even options for configuring the package, may be required in the future.

There is also some discussion at https://github.com/kivy/kivy/issues/5367

###### Things done


- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

vanschelven committed:
Python-kivy: init at 1.10.0

FRidh commented on pkgs/development/python-modules/kivy/default.nix:
```diff
@@ -0,0 +1,99 @@
+buildPythonPackage rec {
+  name = "Kivy-1.10.0";
+
+  src = pkgs.fetchurl {
```
fetchPypi

FRidh commented on pkgs/development/python-modules/kivy/default.nix:
```diff
@@ -0,0 +1,99 @@
+
+  propagatedBuildInputs = with self; [
+    pillow
+    pkgs.mtdev
```
Why is this library propagated?

vanschelven replied:
This is a mistake; I thought that would be helpful, but the actual fix was found later (https://groups.google.com/forum/#!topic/nix-devel/CPQKpM3u1EY), and I never downgraded to non-propagated. Will fix.

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -11176,6 +11176,29 @@ in {
 
+  kivy = callPackage ../development/python-modules/kivy { inherit stdenv buildPythonPackage maintainers platforms licenses pkgs self ; };
+
+  kivygarden = buildPythonPackage rec {
```
in separate file

FRidh commented on pkgs/development/python-modules/kivy/default.nix:
```diff
@@ -0,0 +1,99 @@
+# assert stdenv.lib.versionOlder self.cython.version "0.25.3";
+
+buildPythonPackage rec {
+  name = "Kivy-1.10.0";
```
can you include `pname`, `version`, and `${pname}-${version}` as is now more commonly done.

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -11176,6 +11176,29 @@ in {
+    # The sole reason for this package is that it is a (questionable) dependency of Kivy
+    name = "kivy-garden-0.1.4";
+
+    src = pkgs.fetchurl {
```
fetchPypi

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -11176,6 +11176,29 @@ in {
+
+  kivygarden = buildPythonPackage rec {
+    # The sole reason for this package is that it is a (questionable) dependency of Kivy
+    name = "kivy-garden-0.1.4";
```
same with pname and version

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -11176,6 +11176,29 @@ in {
+      sha256 = "0wkcpr2zc1q5jb0bi7v2dgc0vs5h1y7j42mviyh764j2i0kz8mn2";
+    };
+
+    buildInputs = with self; [ requests ];
```
Is this not needed during runtime? Or is it needed for the tests? In case of the latter, use `checkInputs`.

vanschelven replied:
It's needed during runtime.

kivy-garden is a "not invented here" package manager for kivy components, and as such uses `requests` to fetch those components. 

The real problem is that such a thing exists in the first place, and that it is an explicit dependency of Kivy. I'll talk with the kivy people about both of those issues.

In the meantime, given the fact that the present PR makes kivy-garden available, it should do so in a non-broken way, so I'll promote this to a `propagatedBuildInput`

vanschelven committed:
Kivy: cleanup

As per the comments by @FRidh
(once no more questions/remarks remain, I'll squash the commits into a single one)

FRidh commented on pkgs/development/python-modules/kivy/default.nix:
```diff
@@ -0,0 +1,101 @@
+
+  buildInputs = with self; [
+    pkgconfig
+    cython
```
these two can be `nativeBuildInputs` if I am correct.

vanschelven replied:
I am not confident that I understand the difference (see https://groups.google.com/forum/#!topic/nix-devel/Ad1b0RShe6k) and all other Python packages that I could find do in fact use `buildInputs` here, so I'm a bit wary to proceed with this tip. Feel free to correct me if you're sure about this behavior though.

FRidh commented on pkgs/development/python-modules/kivy/default.nix:
```diff
@@ -0,0 +1,101 @@
+    kivygarden  # See https://github.com/kivy/kivy/issues/5367 suggestion I-2
+    requests  # See https://github.com/kivy/kivy/issues/5367 suggestion I-3
+    docutils # See https://github.com/kivy/kivy/issues/5367 question Q-5
+    pygments # See https://github.com/kivy/kivy/issues/5367 question Q-6
```
Are these 4 only needed during build time? Or during runtime? If the latter -> `propagatedBuildInputs`

vanschelven replied:
Regarding docutils: purely build time.

Regarding the rest: `setup.py` refuses to install without these packages present, but they are fundamentally optional and (in the case of kivygarden especially) it's undesirable to pollute the runtime environment too much with their presence. If someone e.g. want the `pygments` functionality at runtime, it's better to have them install this explicitly than to "pull in the world"

FRidh commented on pkgs/development/python-modules/kivy/default.nix:
```diff
@@ -0,0 +1,101 @@
+    # https://kivy.org/docs/installation/installation.html
+    # https://kivy.org/docs/installation/installation-linux.html
+
+    pkgs.mesa
```
get rid of `pkgs`

FRidh commented on pkgs/development/python-modules/kivy/kivy-garden.nix:
```diff
@@ -0,0 +1,25 @@
+# The sole reason for this package is that it is a (questionable) dependency of Kivy
+{ buildPythonPackage, fetchPypi, maintainers, platforms, licenses, pkgs, self }:
```
same as the other

FRidh commented on pkgs/development/python-modules/kivy/kivy-garden.nix:
```diff
@@ -0,0 +1,25 @@
+    homepage    = "https://pypi.python.org/pypi/kivy-garden";
+    license     = licenses.mit;
+    maintainers = with maintainers; [ vanschelven ];
+    platforms   = platforms.unix;  # Can only test linux; in principle other platforms are supported
```
`buildPythonPackage` already sets platforms

FRidh commented on pkgs/development/python-modules/kivy/default.nix:
```diff
@@ -0,0 +1,101 @@
+    homepage    = "https://pypi.python.org/pypi/kivy";
+    license     = licenses.mit;
+    maintainers = with maintainers; [ vanschelven ];
+    platforms   = platforms.unix;  # I can only test linux; in principle other platforms are supported
```
`buildPythonPackage` already sets platforms

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -11176,6 +11176,10 @@ in {
 
   keyring = callPackage ../development/python-modules/keyring { };
 
+  kivy = callPackage ../development/python-modules/kivy { inherit stdenv buildPythonPackage fetchPypi maintainers platforms licenses pkgs self ; };
```
callPackage will populate the arguments automatically

vanschelven commented on pkgs/development/python-modules/kivy/default.nix:
```diff
@@ -0,0 +1,101 @@
+{ stdenv, buildPythonPackage, fetchPypi, maintainers, platforms, licenses, pkgs, self }:
```
I'm having some trouble implementing this advice, and associated advice on how to remove the associated explicit parameter-setting on the calling side. Could you show me how it's done?

vanschelven commented:
@FRidh I've responded in the above remarking that I'm not confident enough in my knowledge of Nix packaging to push forwards from here. Any chance you could implement your own remarks/requests so that we can push this PR forward and get it accepted?

vanschelven commented:
I'm closing this pull request; it no longer works against the current version of master, and updating is non-trivial at the moment. I might reopen a similar request at a later point in time.
